### PR TITLE
Use quiet-0.2 for some data types

### PR DESCRIPTION
**Pull request by nfrisby**
E.G. `show (SlotNo 3) == "SlotNo 3"` now instead of `"SlotNo {unSlotNo = 3}"`.

### CODE REVIEW

**Comment by erikd** [comment]
Does the `read . show == id` identity still hold?

File: slotting/src/Cardano/Slotting/Block.hs [hunk 1]
```diff
@@ -9,7 +9,8 @@ module Cardano.Slotting.Block
 where
 
 import Cardano.Binary (FromCBOR (..), ToCBOR (..))
-import Cardano.Prelude (NoUnexpectedThunks)
+import Cardano.Prelude (NoUnexpectedThunks, GShowK1 (..),
```
Comment (mrBliss) [inline slotting/src/Cardano/Slotting/Block.hs]: Why is it necessary to import `GShowK1 (..)`? It would be nice if we could avoid that.

Comment (nfrisby) [reply slotting/src/Cardano/Slotting/Block.hs]: Yes, it's an implementation detail of the `GHC.Generics` function, which relies on `coerce`, which needs the constructor "in scope". :(

I _think_ importing the module qualified would satisfy that restriction without having to list `GShowK1` explicitly. Would this be acceptable?

```
import Cardano.Prelude (NoUnexpectedThunks)
import qualified Cardano.Prelude as CPrelude
```

**Comment by nfrisby** [comment]
> Does the `read . show == id` identity still hold?

It does not! I am completely flabbergasted, blindsided. I did not expect this at all. Thank you for explicitly asking the obvious question.

```
Prelude> newtype SlotNo = SlotNo {unSlotNo :: Int} deriving (Read)
Prelude> read "SlotNo 3" :: SlotNo
SlotNo {unSlotNo = *** Exception: Prelude.read: no parse
```

I has assumed `Read` for a record type could also accept the plain arguments. (I.E. That `read . show = id` but also that `read . otherShow` might also be `id`.)

@erikd @mrBliss Should I abandon this path, or should I double-down and also provide a suitable `generic-deriving` `Read` implementation? My hunch is _abandon_: I probably wouldn't have introduced any of this if I had realized `Read` _required_ record syntax.

**Comment by erikd** [comment]
The main reason I use `Read` (which I use less here at IOHK than I have in previous jobs or in my own code) is so that when I get a failure in a property test I can get the counterexample ans paste it into a test file or GHCi to begin debugging. Especially for larger data structures having to edit what I paste to get it accepted is a real pain in the neck.

In order to get around this I take one of two paths:
* Hand implement `Read` and `Show` and use a property test to validate them
* Derive `Read` and `Show` and have a separate `renderX` function that does pretty printing.

Like you said, here at IOHK we tend not to use `Read` much, so this may not be an issue.

**Comment by nfrisby** [comment]
> The main reason I use `Read` (which I use less here at IOHK than I have in previous jobs or in my own code) is so that when I get a failure in a property test I can get the counterexample ans paste it into a test file or GHCi to begin debugging. Especially for larger data structures having to edit what I paste to get it accepted is a real pain in the neck.

I absolutely feel that same pain! I have been copying-and-pasting data structures from QuickCheck output that contain multiple `Map`/`Set`s multiples times per day, and qualifying each `fromList` is obnoxious. I didn't think to use `read` for that (it's so far off my radar); it's a great use in the short-term.

In fact, the main motivation for this whole `gshowsPrecWithoutRecordSyntax` is for the QuickCheck output etc to be more legible in the first place :)

**Comment by erikd** [comment]
> n fact, the main motivation for this whole gshowsPrecWithoutRecordSyntax is for the QuickCheck output etc to be more legible in the first place :)

But standard derived `Read`/`Show` instances may be the easiest way to get that result.

**Comment by erikd** [comment]
@nfrisby https://hackage.haskell.org/package/quiet-0.2/docs/Quiet.html

Is that of any use?

**Comment by nfrisby** [comment]
> But standard derived Read/Show instances may be the easiest way to get that result.

Unless you mean something else by "standard", I find the record syntax on these simplistic constructors makes it quite hard to read the expression.

> @nfrisby https://hackage.haskell.org/package/quiet-0.2/docs/Quiet.html
> 
> Is that of any use?

Yes, that looks like the exact sort of thing I wanted. I'm weary that it doesn't rely on `generic-deriving`'s definitions of `read/show`, which I consider to be the most mature behind GHC's, but getting `Read/Show` right isn't Herculean after all and `Quiet`'s test suite looks non-trivial. Thanks for the pointer.

**Comment by erikd** [comment]
@nfrisby How would `generic-sop` improve `quiet` ?

**Comment by jacobstanley** [comment]
> I'm weary that it doesn't rely on `generic-deriving`'s definitions of `read/show`

@nfrisby sorry to barge in, I was talking with @erikd offline

I have been carrying around the `quiet` code for a number of years, only put it on Hackage this week, but if I remember right the `Show` code is a [derivative](https://github.com/jacobstanley/quiet/blob/master/LICENSE#L2) of `generic-deriving`. The `Read` code I wrote myself, but as you note is fairly comprehensively tested.

The code is simpler than `generic-deriving` as it defers to `Show` for everything but the generic structure (rather that implementing a new `GShow` typeclass for all types)

I use it for the same pains discussed here, QuickCheck/Hedgehog counterexamples!

**Comment by edsko** [comment]
I think copy-and-paste-ability into ghci is indeed very useful. Just want to point out that the derived `Show` without the record fields _still_ satisfies that requirement.

**Comment by erikd** [comment]
@edsko I do not think that is true when the constructor is not exported from the module it is defined in (eg for a smart constructor).

**Comment by edsko** [comment]
> @edsko I do not think that is true when the constructor is not exported from the module it is defined in (eg in order to force use of an alternative smart constructor).

I didn't articulate myself very well. You are right of course, a derived `Show` instance versus a custom one might not work. I was commenting merely on @nfrisby 's changes here: _if_ the standard derived `Show` instance works, then _his_ derived `Show` instance (which omits the record names) would _also_ still work.

**Comment by nfrisby** [comment]
> @nfrisby sorry to barge in, I was talking with @erikd offline

Not at all -- thanks for commenting! And thank you for packaging up and sharing your work on Hackage.

**Comment by nfrisby** [comment]
OK. I've updated this to use `quiet`, but it will need input-output-hk/cardano-prelude#111, which adds `quiet` to the wrapped LTS snapshot. Once that merges, I can update the build config files here accordingly.

File: slotting/src/Cardano/Slotting/Block.hs [hunk 2]
```diff
@@ -12,14 +12,18 @@ import Cardano.Prelude (NoUnexpectedThunks)
...
 -- | The 0-based index of the block in the blockchain.
 -- BlockNo is <= SlotNo and is only equal at slot N if there is a block
 -- for every slot where N <= SlotNo.
 newtype BlockNo = BlockNo {unBlockNo :: Word64}
-  deriving stock (Show, Eq, Ord, Generic)
+  deriving stock (Eq, Ord, Generic)
   deriving newtype (Enum, Bounded, Num, Serialise, NoUnexpectedThunks)
 
+instance Show BlockNo where
+  showsPrec = Quiet.qshowsPrec
```
Comment (mrBliss) [inline slotting/src/Cardano/Slotting/Block.hs:25]: I would have used
```haskell
import Quiet (Quiet (..))

  ..
  deriving (Show) via (Quiet BlockNo)
```
Not a fan?

Comment (nfrisby) [reply slotting/src/Cardano/Slotting/Block.hs]: Same response as https://github.com/input-output-hk/ouroboros-network/pull/1019#discussion_r427992547

**Comment by newhoggy** [comment]
Looks like this is dealt with by the related PR https://github.com/input-output-hk/cardano-base/pull/151

Closing.

**Pull request closed without merging**